File: Modeling/app/src/main/java/pisarev/com/modeling/mvp/view/customview/DrawView.java
package pisarev.com.modeling.mvp.view.customview;

import android.app.AlertDialog;
import android.content.Context;
import android.content.SharedPreferences;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.DashPathEffect;
import android.graphics.Paint;
import android.graphics.Path;
import android.os.Build;
import android.preference.PreferenceManager;
import android.support.annotation.Nullable;
import android.support.annotation.RequiresApi;
import android.util.AttributeSet;
import android.util.Log;
import android.view.MotionEvent;
import android.view.ScaleGestureDetector;
import android.view.View;

import java.util.*;

import pisarev.com.modeling.interfaces.DrawMvp;
import pisarev.com.modeling.interfaces.Drawing;
import pisarev.com.modeling.interfaces.IDraw;
import pisarev.com.modeling.mvp.model.*;

import static android.view.MotionEvent.*;

public class DrawView extends View implements IDraw, DrawMvp.PresenterDrawViewMvp {
    private Paint paintCoordinateDottedLine;
    private DrawMvp.DrawViewMvp drawActivity;
    private Point pointSystemCoordinate;
    private boolean isTouch;
    private boolean isMove;
    private float moveX;
    private float moveZ;
    private float zooming = 3;
    public int button;
    public final int START = 1;
    public final int RESET = 2;
    public final int STOP = 3;
    public final int TAP = 4;
    private int index;
    private int numberLine;
    private ScaleGestureDetector scaleGestureDetector;
    private ArrayList<String> errorList;
    private boolean isSingleBlockDown = false;
    private boolean isResetDown = false;
    private boolean isStartDown = false;
    private final String TEG = getClass().getName();
    private MyData data;
    private Drawing drawing;
    private boolean isDrawPoint = false;

    public DrawView(Context context) {
        super(context);
        init(context);
    }

    public DrawView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init(context);
    }

    private void init(Context context) {
        scaleGestureDetector = new ScaleGestureDetector(context, new ScaleListener());
        SharedPreferences myPreferences = PreferenceManager.getDefaultSharedPreferences(context);
        data = new MyData();
        paintCoordinateDottedLine = new Paint();
        paintCoordinateDottedLine.setColor(Color.LTGRAY);
        paintCoordinateDottedLine.setStyle(Paint.Style.STROKE);
        paintCoordinateDottedLine.setAntiAlias(true);
        paintCoordinateDottedLine.setPathEffect(new DashPathEffect(new float[]{20f, 10f}, 0f));
        pointSystemCoordinate = new Point();
        errorList = new ArrayList<>();
        if (myPreferences.getBoolean("RADIOBUTTON", false)) {
            drawing = new DrawVerticalTurning(this);
        } else {
            drawing = new DrawHorizontalTurning(this);
        }
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        switch (button) {
            case START:
            case STOP:
                drawing.drawContour(canvas,data, pointSystemCoordinate, zooming, index);
                invalidate();
                break;
            case TAP:
                drawing.setNumberLine(numberLine);
                drawing.drawContour(canvas,data, pointSystemCoordinate, zooming, index);
                invalidate();
                break;
            case RESET:
                initSystemCoordinate(canvas, true);
                button = 0;
                invalidate();
                isTouch = false;
                drawing.setNumberLine(-1);
                errorList.clear();
                break;
        }
        drawSystemCoordinate(canvas, isTouch);
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    @Override
    public boolean onTouchEvent(MotionEvent event) {
        scaleGestureDetector.onTouchEvent(event);
        switch (event.getAction()) {
            case ACTION_DOWN:
                float downX = event.getX();
                float downZ = event.getY();
                moveX = pointSystemCoordinate.getX() - downX;
                moveZ = pointSystemCoordinate.getZ() - downZ;
                isTouch = true;
                invalidate();
                break;
            case ACTION_MOVE:
                pointSystemCoordinate.setX(event.getX() + moveX);
                pointSystemCoordinate.setZ(event.getY() + moveZ);
                invalidate();
                isMove=true;
                break;
            case ACTION_UP:
                if(!isMove){
                    Point point = new Point();
                    point.setX((pointSystemCoordinate.getX() - event.getX()) * -1);
                    point.setZ(event.getY());
                    if (point.getZ() > 0) point.setZ(pointSystemCoordinate.getZ() - point.getZ());
                    else point.setZ(pointSystemCoordinate.getZ() + Math.abs(point.getZ()));
                    point.setX(point.getX() / zooming);
                    point.setZ(point.getZ() / zooming);
                    Optional<Frame> frame = getFrame(point);
                    if (frame.isPresent()) {
                        numberLine=frame.get().getId();
                        button=TAP;
                        invalidate();
                        drawActivity.showFrame(data.getProgramList().get(frame.get().getId()).toString());
                        drawActivity.showAxis("X=" + frame.get().getX(), "Z=" + frame.get().getZ());
                        System.out.println("X="+point.getX());
                        System.out.println("Z="+point.getZ());
                        System.out.println(data.getProgramList().get(frame.get().getId()));
                        isDrawPoint = true;
                    } else if (isDrawPoint) {
                        numberLine=-1;
                        button=TAP;
                        drawActivity.showFrame("");
                        drawActivity.showAxis("", "");
                        isDrawPoint = false;
                    }
                }
                isMove=false;
                invalidate();
                break;
        }
        return true;
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    private Optional<Frame> getFrame(Point point) {
        int side = 20;
        Rect rect = new Rect();
        rect.setRect(point.getX() - (side >> 1), point.getZ() - (side >> 1), side, side);
        return data.getFrameList().stream()
                .filter(p -> rect.isInsideRect(p.getX(), p.getZ()))
                .min(Comparator.comparingDouble(p -> Math.abs(point.getX() - p.getX()) + Math.abs(point.getZ() - p.getZ())));
    }

    @Override
    public void showError(String error) {
        button = STOP;
        if (!errorList.contains(error)) {
            errorList.add(error);
            AlertDialog.Builder builder = new AlertDialog.Builder(getContext());
            builder.setTitle("Error")
                    .setMessage(error)
                    .setCancelable(false)
                    .setNegativeButton("OK", (dialog, which) -> dialog.cancel());
            AlertDialog alertDialog = builder.create();
            alertDialog.show();
            Log.d(TEG, "error " + error);
        }
    }


    private void drawSystemCoordinate(Canvas canvas, boolean isTouch) {
        if (!isTouch || button == RESET) {
            initSystemCoordinate(canvas, true);
            invalidate();
        }
        if (isTouch || button == START) {
            initSystemCoordinate(canvas, false);
            drawing.drawContour(canvas,data, pointSystemCoordinate, zooming, index);
            invalidate();
        }

    }

    private void initSystemCoordinate(Canvas canvas, boolean isInit) {
        Path path;
        if (isInit) {
            path = new Path();
            pointSystemCoordinate.setX(getWidth() >> 1);
            pointSystemCoordinate.setZ(getHeight() >> 1);
            path.moveTo(0, pointSystemCoordinate.getZ());
            path.lineTo(getWidth(), pointSystemCoordinate.getZ());
            path.moveTo(pointSystemCoordinate.getX(), 0);
            path.lineTo(pointSystemCoordinate.getX(), getHeight());
            canvas.drawPath(path, paintCoordinateDottedLine);
        } else {
            path = new Path();
            path.moveTo(0, pointSystemCoordinate.getZ());
            path.lineTo(getWidth(), pointSystemCoordinate.getZ());
            path.moveTo(pointSystemCoordinate.getX(), 0);
            path.lineTo(pointSystemCoordinate.getX(), getHeight());
            canvas.drawPath(path, paintCoordinateDottedLine);
        }
    }

    @Override
    public void onButtonStart() {
        isStartDown = true;
        index = data.getFrameList().size();
        if (data.getFrameList().size() > 0) {
            button = START;
            isResetDown = false;
        }
    }

    @Override
    public void onButtonCycleStart() {
        if (isSingleBlockDown && index < data.getFrameList().size()) {
            isResetDown = false;
            button = START;
            index++;
            drawActivity.showFrame((data.getProgramList().get(data.getFrameList().get(index - 1).getId())).toString());
            if (data.getFrameList().get(index - 1).isAxisContains()) {
                drawActivity.showAxis("X=" + data.getFrameList().get(index - 1).getX(), "Z=" + data.getFrameList().get(index - 1).getZ());
                drawActivity.showIndex(index);
            }
        }
        if (!isSingleBlockDown && index < data.getFrameList().size() && !isStartDown) {
            isResetDown = false;
            isStartDown = true;
            button = START;
            final Timer timer = new Timer();
            timer.schedule(new TimerTask() {
                @Override
                public void run() {
                    if (index < data.getFrameList().size() && !isSingleBlockDown && !isResetDown && button == START) {
                        index++;
                        drawActivity.showFrame((data.getProgramList().get(data.getFrameList().get(index - 1).getId())).toString());
                        if (data.getFrameList().get(index - 1).isAxisContains()) {
                            drawActivity.showAxis("X=" + data.getFrameList().get(index - 1).getX(), "Z=" + data.getFrameList().get(index - 1).getZ());
                            drawActivity.showIndex(index);
                        }
                    } else {
                        isResetDown = false;
                        timer.cancel();
                        if (button == STOP) {
                            drawActivity.showFrame((data.getProgramList().get(data.getFrameList().get(index - 1).getId())).toString());
                            if (data.getFrameList().get(index - 1).isAxisContains()) {
                                drawActivity.showAxis("X=" + data.getFrameList().get(index - 1).getX(), "Z=" + data.getFrameList().get(index - 1).getZ());
                                drawActivity.showIndex(index);
                            }
                        }
                    }
                }
            }, 0, 200);
        }
    }

    @Override
    public void onButtonSingleBlock(boolean isSingleBlockDown) {
        this.isSingleBlockDown = isSingleBlockDown;
        isStartDown = false;
    }

    @Override
    public void onButtonReset() {
        isResetDown = true;
        button = RESET;
        index = 0;
        isStartDown = false;
        drawActivity.showFrame("");
        drawActivity.showAxis("", "");
    }

    @Override
    public void setActivity(DrawMvp.DrawViewMvp secondView) {
        this.drawActivity = secondView;
    }

    @Override
    public void setIndex(int index) {
        this.index = index;
    }

    @Override
    public void setData(MyData data) {
        this.data = data;
    }

    public class ScaleListener extends ScaleGestureDetector.SimpleOnScaleGestureListener {

        @Override
        public boolean onScale(ScaleGestureDetector detector) {
            zooming *= detector.getScaleFactor();
            invalidate();
            return true;
        }
    }
}

File: Modeling/app/src/main/java/pisarev/com/modeling/mvp/view/customview/DrawView.kt
package pisarev.com.modeling.mvp.view.customview

import pisarev.com.modeling.interfaces.IDraw
import pisarev.com.modeling.interfaces.DrawMvp.PresenterDrawViewMvp
import pisarev.com.modeling.interfaces.DrawMvp.DrawViewMvp
import android.view.ScaleGestureDetector
import pisarev.com.modeling.mvp.model.core.MyData
import pisarev.com.modeling.interfaces.Drawing
import android.preference.PreferenceManager
import pisarev.com.modeling.mvp.contur.DrawVerticalTurning
import android.annotation.SuppressLint
import android.app.AlertDialog
import android.content.Context
import android.support.annotation.RequiresApi
import android.os.Build
import android.view.MotionEvent
import android.content.DialogInterface
import android.graphics.*
import android.util.AttributeSet
import android.util.Log
import android.view.ScaleGestureDetector.SimpleOnScaleGestureListener
import android.view.View
import pisarev.com.modeling.mvp.model.core.Frame
import pisarev.com.modeling.mvp.contur.Point
import pisarev.com.modeling.mvp.model.Rect
import java.util.*
import kotlin.math.abs

class DrawView : View, IDraw, PresenterDrawViewMvp {
    private var paintCoordinateDottedLine: Paint? = null
    private var drawActivity: DrawViewMvp? = null
    private var pointSystemCoordinate: Point? = null
    private var isTouch = false
    private var isMove = false
    private var moveX = 0f
    private var moveZ = 0f
    private var zooming = 3f
    var button = 0
    val START = 1
    val RESET = 2
    val STOP = 3
    val TAP = 4
    private var index = 0
    private var numberLine = 0
    private var scaleGestureDetector: ScaleGestureDetector? = null
    private var errorList: ArrayList<String>? = null
    private var isSingleBlockDown = false
    private var isResetDown = false
    private var isStartDown = false
    private val TEG = javaClass.name
    private var data: MyData? = null
    private var drawing: Drawing? = null
    private var isDrawPoint = false
    private var pointStopCanvas: Point? = null

    constructor(context: Context) : super(context) {
        init(context)
    }

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        init(context)
    }

    private fun init(context: Context) {
        scaleGestureDetector = ScaleGestureDetector(context, ScaleListener())
        val myPreferences = PreferenceManager.getDefaultSharedPreferences(context)
        data = MyData()
        paintCoordinateDottedLine = Paint()
        paintCoordinateDottedLine!!.color = Color.LTGRAY
        paintCoordinateDottedLine!!.style = Paint.Style.STROKE
        paintCoordinateDottedLine!!.isAntiAlias = true
        paintCoordinateDottedLine!!.pathEffect = DashPathEffect(floatArrayOf(20f, 10f), 0f)
        pointSystemCoordinate = Point()
        errorList = ArrayList()
        drawing = DrawVerticalTurning(this)
    }

    @SuppressLint("DrawAllocation")
    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        when (button) {
            START, STOP -> {
                drawing!!.drawContour(canvas, data!!, Point(pointSystemCoordinate!!), zooming, index)
                invalidate()
            }
            TAP -> {
                drawing!!.setNumberLine(numberLine)
                drawing!!.drawContour(canvas, data!!, Point(pointSystemCoordinate!!), zooming, index)
                invalidate()
            }
            RESET -> {
                button = 0
                invalidate()
                drawing!!.setNumberLine(-1)
                errorList!!.clear()
            }
        }
        drawSystemCoordinate(canvas, isTouch)
    }

    @SuppressLint("ClickableViewAccessibility")
    @RequiresApi(api = Build.VERSION_CODES.N)
    override fun onTouchEvent(event: MotionEvent): Boolean {
        scaleGestureDetector!!.onTouchEvent(event)
        when (event.action) {
            MotionEvent.ACTION_DOWN -> {
                val downX = event.x
                val downZ = event.y
                moveX = pointSystemCoordinate!!.x - downX
                moveZ = pointSystemCoordinate!!.z - downZ
                isTouch = true
                invalidate()
            }
            MotionEvent.ACTION_MOVE -> {
                pointSystemCoordinate!!.x=event.x + moveX
                pointSystemCoordinate!!.z=event.y + moveZ
                invalidate()
                isMove = true
            }
            MotionEvent.ACTION_UP -> {
                if (!isMove) {
                    val point = Point()
                    point.x=(pointSystemCoordinate!!.x - event.x) * -1
                    point.z=event.y
                    if (point.z > 0) point.z=pointSystemCoordinate!!.z - point.z else point.z= pointSystemCoordinate!!.z + Math.abs(point.z)
                    point.x=point.x / zooming
                    point.z=point.z/ zooming
                    val frame = getFrame(point)
                    if (frame.isPresent) {
                        numberLine = frame.get().id
                        button = TAP
                        invalidate()
                        drawActivity!!.showFrame(data!!.programList[frame.get().id].toString())
                        drawActivity!!.showAxis("X=" + frame.get().x, "Z=" + frame.get().z)
                        System.out.println("X=" + point.x)
                        System.out.println("Z=" + point.z)
                        println(data!!.programList[frame.get().id])
                        isDrawPoint = true
                    } else if (isDrawPoint) {
                        numberLine = -1
                        button = TAP
                        drawActivity!!.showFrame("")
                        drawActivity!!.showAxis("", "")
                        isDrawPoint = false
                    }
                }
                isMove = false
                invalidate()
            }
        }
        return true
    }

    @RequiresApi(api = Build.VERSION_CODES.N)
    private fun getFrame(point: Point): Optional<Frame> {
        val side = 20
        val rect = Rect()
        rect.setRect(point.x - (side shr 1), point.z - (side shr 1), side.toFloat(), side.toFloat())
        return data!!.frameList.stream()
            .filter { rect.isInsideRect(point) }
            .min(Comparator.comparingDouble { p: Frame -> (abs(point.x - p.x) + abs(point.z - p.z)) })
    }

    override fun showError(error: String) {
        button = STOP
        if (!errorList!!.contains(error)) {
            errorList!!.add(error)
            val builder = AlertDialog.Builder(context)
            builder.setTitle("Error")
                .setMessage(error)
                .setCancelable(false)
                .setNegativeButton("OK") { dialog: DialogInterface, which: Int -> dialog.cancel() }
            val alertDialog = builder.create()
            alertDialog.show()
            Log.d(TEG, "error $error")
        }
    }

    private fun drawSystemCoordinate(canvas: Canvas, isTouch: Boolean) {
        if (!isTouch || button == RESET) {
            initSystemCoordinate(canvas, true)
            invalidate()
        }
        if (isTouch || button == START) {
            drawing!!.drawContour(canvas, data!!, Point(pointSystemCoordinate!!), zooming, index)
            initSystemCoordinate(canvas, false)
            invalidate()
        }
    }

    private fun initSystemCoordinate(canvas: Canvas, isInit: Boolean) {
        val path: Path
        if (isInit) {
            path = Path()
            pointSystemCoordinate!!.x= (width shr 1).toFloat()
            pointSystemCoordinate!!.z=(height shr 1).toFloat()
            path.moveTo(0f, pointSystemCoordinate!!.z)
            path.lineTo(width.toFloat(), pointSystemCoordinate!!.z)
            path.moveTo(pointSystemCoordinate!!.x, 0f)
            path.lineTo(pointSystemCoordinate!!.x, height.toFloat())
            canvas.drawPath(path, paintCoordinateDottedLine!!)
        } else {
            path = Path()
            path.moveTo(0f, pointSystemCoordinate!!.z)
            path.lineTo(width.toFloat(), pointSystemCoordinate!!.z)
            path.moveTo(pointSystemCoordinate!!.x, 0f)
            path.lineTo(pointSystemCoordinate!!.x, height.toFloat())
            canvas.drawPath(path, paintCoordinateDottedLine!!)
        }
    }

    override fun onButtonStart() {
        isStartDown = true
        index = data!!.frameList.size
        if (data!!.frameList.size > 0) {
            button = START
            isResetDown = false
        }
    }

    override fun onButtonCycleStart() {
        if (isSingleBlockDown && index < data!!.frameList.size) {
            isResetDown = false
            button = START
            index++
            drawActivity!!.showFrame(data!!.programList[data!!.frameList[index - 1].id].toString())
            if (data!!.frameList[index - 1].isAxisContains) {
                drawActivity!!.showAxis("X=" + data!!.frameList[index - 1].x, "Z=" + data!!.frameList[index - 1].z)
                drawActivity!!.showIndex(index)
            }
        }
        if (!isSingleBlockDown && index < data!!.frameList.size && !isStartDown) {
            isResetDown = false
            isStartDown = true
            button = START
            val timer = Timer()
            timer.schedule(object : TimerTask() {
                override fun run() {
                    if (index < data!!.frameList.size && !isSingleBlockDown && !isResetDown && button == START) {
                        index++
                        drawActivity!!.showFrame(data!!.programList[data!!.frameList[index - 1].id].toString())
                        if (data!!.frameList[index - 1].isAxisContains) {
                            drawActivity!!.showAxis(
                                "X=" + data!!.frameList[index - 1].x,
                                "Z=" + data!!.frameList[index - 1].z
                            )
                            drawActivity!!.showIndex(index)
                        }
                    } else {
                        isResetDown = false
                        timer.cancel()
                        if (button == STOP) {
                            drawActivity!!.showFrame(data!!.programList[data!!.frameList[index - 1].id].toString())
                            if (data!!.frameList[index - 1].isAxisContains) {
                                drawActivity!!.showAxis(
                                    "X=" + data!!.frameList[index - 1].x,
                                    "Z=" + data!!.frameList[index - 1].z
                                )
                                drawActivity!!.showIndex(index)
                            }
                        }
                    }
                }
            }, 0, 200)
        }
    }

    override fun onButtonSingleBlock(isSingleBlockDown: Boolean) {
        this.isSingleBlockDown = isSingleBlockDown
        isStartDown = false
    }

    override fun onButtonReset() {
        isResetDown = true
        button = RESET
        index = 0
        isStartDown = false
        drawActivity!!.showFrame("")
        drawActivity!!.showAxis("", "")
    }

    override fun setActivity(secondView: DrawViewMvp) {
        drawActivity = secondView
    }

    override fun setIndex(index: Int) {
        this.index = index
    }

    override fun setData(data: MyData) {
        this.data = data
    }

    inner class ScaleListener : SimpleOnScaleGestureListener() {
        override fun onScale(detector: ScaleGestureDetector): Boolean {
            zooming *= detector.scaleFactor
            val point = Point()
            point.x=(pointSystemCoordinate!!.x - detector.focusX) * -1
            point.z=detector.focusY
            if (point.z > 0) point.z=pointSystemCoordinate!!.z - point.z else point.z=
                pointSystemCoordinate!!.z + Math.abs(point.z
            )
            pointStopCanvas = point
            point.x=pointStopCanvas!!.x * zooming
            point.z=pointStopCanvas!!.z * zooming
            pointSystemCoordinate!!.x=pointSystemCoordinate!!.x + pointStopCanvas!!.x - point.x
            pointSystemCoordinate!!.z=pointSystemCoordinate!!.z + point.z - pointStopCanvas!!.z
            pointStopCanvas!!.x=pointStopCanvas!!.x * zooming
            pointStopCanvas!!.z=pointStopCanvas!!.z * zooming
            return true
        }
    }
}